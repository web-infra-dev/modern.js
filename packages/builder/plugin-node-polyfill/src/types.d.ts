declare module 'node-libs-browser';
