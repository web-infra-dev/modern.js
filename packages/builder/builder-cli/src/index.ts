export { defineConfig } from '@modern-js/builder/cli';
