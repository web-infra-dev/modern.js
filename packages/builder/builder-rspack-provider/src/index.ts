export { getRspackVersion } from './shared/rspackVersion';
export { builderRspackProvider } from './provider';
export type { BuilderRspackProvider } from './provider';

export type {
  // Config Types
  BuilderConfig,
  NormalizedConfig,

  // Hook Callback Types
  ModifyRspackConfigFn,

  // Plugin Types
  BuilderPluginAPI,

  // Rspack
  Rspack,
  RspackConfig,
} from './types';
