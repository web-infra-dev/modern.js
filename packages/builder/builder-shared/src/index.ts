export * from './applyDefaultBuilderOptions';
export * from './constants';
export * from './createHook';
export * from './pluginStore';
export * from './createContext';
export * from './utils';
export * from './fs';
export * from './generateMetaTags';
export * from './getBrowserslist';
export * from './getCssSupport';
export * from './logger';
export * from './mergeBuilderConfig';
export * from './onExitProcess';
export * from './pick';
export * from './regexp';
export * from './types';
export * from './config';
export * from './format';
export * from './devServer';
export * from './prodServer';
export * from './chain';
export * from './schema';
export * from './plugins';
export * from './openBrowser';
export * from './apply';
export * from './fallback';
export * from './getLoaderOptions';
export * from './svgo';
export * from './patch';
export * from './manifest';
export * from './css';
export * from './minimize';
export * from './core-js';
export * from './progress';
export * from './prettyTime';
