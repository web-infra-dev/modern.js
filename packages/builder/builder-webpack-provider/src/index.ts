export { builderWebpackProvider } from './provider';
export type { BuilderWebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export type { WebpackBuildError } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  BuilderPlugin,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CSSExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
