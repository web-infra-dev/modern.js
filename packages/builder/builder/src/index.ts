export { createBuilder } from './createBuilder';
export { mergeBuilderConfig } from '@modern-js/builder-shared';

export { builderPluginSourceBuild } from './plugins/sourceBuild';

export type {
  BuilderMode,
  BuilderEntry,
  BuilderTarget,
  BuilderPlugin,
  BuilderContext,
  BuilderInstance,
  CreateBuilderOptions,
  InspectConfigOptions,

  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeCreateCompilerFn,
  OnDevCompileDoneFn,
  ModifyBuilderConfigFn,
} from '@modern-js/builder-shared';
