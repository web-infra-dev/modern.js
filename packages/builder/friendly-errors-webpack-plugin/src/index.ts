export * from './shared/types';
export * from './shared/utils';
export * from './plugin';
