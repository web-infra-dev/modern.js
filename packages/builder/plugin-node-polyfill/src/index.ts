import type { BuilderPlugin } from '@modern-js/webpack-builder';

/**
 * Usage:
 *
 *   const { PluginNodePolyfill } = await import('@modern-js/webpack-builder-plugin-node-polyfill');
 *
 *   builder.addPlugins([ PluginNodePolyfill() ]);
 */
export function PluginNodePolyfill(): BuilderPlugin {
  return {
    name: 'webpack-builder-plugin-node-polyfill',

    async setup(api) {
      const { default: webpack } = await import('webpack');

      const { default: nodeLibsBrowser } = await import('node-libs-browser');

      api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
        chain
          .plugin(CHAIN_ID.PLUGIN.NODE_POLYFILL_PROVIDE)
          .use(webpack.ProvidePlugin, [
            {
              Buffer: [nodeLibsBrowser.buffer, 'Buffer'],
              process: [nodeLibsBrowser.process],
            },
          ]);

        const fallback = Object.keys(nodeLibsBrowser).reduce<
          Record<string, string | false>
        >((previous, name) => {
          if (nodeLibsBrowser[name]) {
            previous[name] = nodeLibsBrowser[name];
          } else {
            previous[name] = false;
          }
          return previous;
        }, {});

        chain.resolve.fallback.merge(fallback);
      });
    },
  };
}
