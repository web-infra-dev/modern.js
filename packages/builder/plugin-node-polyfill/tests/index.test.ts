import { expect, describe, it } from 'vitest';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { builderPluginNodePolyfill } from '../src';

describe('plugins/node-polyfill', () => {
  it('should add node-polyfill config', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginNodePolyfill()],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
