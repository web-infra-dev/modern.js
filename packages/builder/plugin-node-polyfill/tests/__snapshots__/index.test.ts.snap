// Vitest Snapshot v1

exports[`plugins/node-polyfill > should add node-polyfill config 1`] = `
{
  "plugins": [
    ProvidePlugin {
      "definitions": {
        "Buffer": [
          "<ROOT>/node_modules/<PNPM_INNER>/buffer/index.js",
          "Buffer",
        ],
        "process": [
          "<ROOT>/node_modules/<PNPM_INNER>/process/browser.js",
        ],
      },
    },
  ],
  "resolve": {
    "fallback": {
      "_stream_duplex": "<ROOT>/node_modules/<PNPM_INNER>/readable-stream/duplex.js",
      "_stream_passthrough": "<ROOT>/node_modules/<PNPM_INNER>/readable-stream/passthrough.js",
      "_stream_readable": "<ROOT>/node_modules/<PNPM_INNER>/readable-stream/readable.js",
      "_stream_transform": "<ROOT>/node_modules/<PNPM_INNER>/readable-stream/transform.js",
      "_stream_writable": "<ROOT>/node_modules/<PNPM_INNER>/readable-stream/writable.js",
      "assert": "<ROOT>/node_modules/<PNPM_INNER>/assert/assert.js",
      "buffer": "<ROOT>/node_modules/<PNPM_INNER>/buffer/index.js",
      "child_process": false,
      "cluster": false,
      "console": "<ROOT>/node_modules/<PNPM_INNER>/console-browserify/index.js",
      "constants": "<ROOT>/node_modules/<PNPM_INNER>/constants-browserify/constants.json",
      "crypto": "<ROOT>/node_modules/<PNPM_INNER>/crypto-browserify/index.js",
      "dgram": false,
      "dns": false,
      "domain": "<ROOT>/node_modules/<PNPM_INNER>/domain-browser/source/index.js",
      "events": "<ROOT>/node_modules/<PNPM_INNER>/events/events.js",
      "fs": false,
      "http": "<ROOT>/node_modules/<PNPM_INNER>/stream-http/index.js",
      "https": "<ROOT>/node_modules/<PNPM_INNER>/https-browserify/index.js",
      "module": false,
      "net": false,
      "os": "<ROOT>/node_modules/<PNPM_INNER>/os-browserify/browser.js",
      "path": "<ROOT>/node_modules/<PNPM_INNER>/path-browserify/index.js",
      "process": "<ROOT>/node_modules/<PNPM_INNER>/process/browser.js",
      "punycode": "<ROOT>/node_modules/<PNPM_INNER>/punycode/punycode.js",
      "querystring": "<ROOT>/node_modules/<PNPM_INNER>/querystring-es3/index.js",
      "readline": false,
      "repl": false,
      "stream": "<ROOT>/node_modules/<PNPM_INNER>/stream-browserify/index.js",
      "string_decoder": "<ROOT>/node_modules/<PNPM_INNER>/string_decoder/lib/string_decoder.js",
      "sys": "<ROOT>/node_modules/<PNPM_INNER>/util/util.js",
      "timers": "<ROOT>/node_modules/<PNPM_INNER>/timers-browserify/main.js",
      "tls": false,
      "tty": "<ROOT>/node_modules/<PNPM_INNER>/tty-browserify/index.js",
      "url": "<ROOT>/node_modules/<PNPM_INNER>/url/url.js",
      "util": "<ROOT>/node_modules/<PNPM_INNER>/util/util.js",
      "vm": "<ROOT>/node_modules/<PNPM_INNER>/vm-browserify/index.js",
      "zlib": "<ROOT>/node_modules/<PNPM_INNER>/browserify-zlib/lib/index.js",
    },
  },
}
`;
